Web/Controller/ImageStyle.hs

```diff
@@ -0,0 +1,47 @@
+module Web.Controller.ImageStyle where
+
+import Web.Controller.Prelude
+import IHP.ControllerSupport
+import System.Directory (doesFileExist)
+import qualified Data.Text as T
+
+
+instance Controller ImageStyleController where
+    action RenderImageStyleAction { width, height, originalImagePath } = do
+        -- Remove the prefix, and extract the UUID.
+        let (originalImageDirectory, uuid) = extractDirectoryAndUUID originalImagePath
+
+        let size = show width <> "x" <> show height
+
+        let imageStylePathDirectory = originalImageDirectory <> "/imageStyles/" <> size
+        let imageStylePath = imageStylePathDirectory <> "/" <> uuid
+
+
+        -- @todo: How to get rid of the "static/" prefix?
+        fileExists <- doesFileExist (cs $ "static/" <> imageStylePath)
+
+        if fileExists
+            then do
+                -- Image style found.
+                renderFile (cs $ "static/" <> imageStylePath) "application/jpg"
+            else do
+                -- Image style not found, so create it.
+                let options :: StoreFileOptions = def
+                        { directory = imageStylePathDirectory
+                        , preprocess = applyImageMagick "jpg" ["-resize", cs size <> "^", "-gravity", "north", "-crop", cs size, "-quality", "85%", "-strip"]
+                        }
+
+                storedFile <- storeFileFromPath (cs $ "static/" <> originalImageDirectory <> "/" <> uuid) options
```

Comment:
Need to remove the  hardcoded `static/`, but the bigger issue is this

![image](https://github.com/amitaibu/ihp-landing-page/assets/125707/03100e19-37e6-49b7-b4c3-6e70712aabfc)

Reply:
The error is triggered when we have the `preprocess` from above

Reply:
I've fixed the imageMagick params https://github.com/amitaibu/ihp-landing-page/pull/15/commits/5c0c86e2a7b913d804c663ccc843be54d6be56e5

Reply:
Next problem, how to set the filename on `storeFileFromPath` -- right now it creates duplicate files on each reload 

![image](https://github.com/amitaibu/ihp-landing-page/assets/125707/c55a5d26-9a51-4be4-bfd9-a4fd4593e8ff)

Reply:
https://github.com/digitallyinduced/ihp/blob/248cb16b361adb4e8838b33a496f7c2be09541de/IHP/FileStorage/ControllerFunctions.hs#L181

@mpscholten is there a good reason for why the filename is `""`? Should I write a PR to change it to take the same filename as the original file?

Reply:
https://github.com/digitallyinduced/ihp/pull/1686

Reply:
For prod usage it's btw not recommended to use the local storage. Better to use e.g. the S3 storage. The problem for production is that on NixOS the static/ directory is not writeable. So keep this in mind while desgining this controller.

Reply:
> security reasons

Indeed, it makes the file name secure :+1: . However for my use case I need the storeFileFromPath to save the same filename, so I won't re-create the image style each time with a different name.So https://github.com/digitallyinduced/ihp/pull/1686 might not be right as is, but I think there should be a way to tell IHP the filename to save

>  Then there should be a database record that saves the storage path together with the original file name (as a string), e.g. in an attachments table.

Nice. I could add it to the docs :) . At which point do you get a hold of both filename and path?

Reply:
You could work around this by accessing the underlying `StaticDirStorage` or `S3Storage` directly, but the storage system right now is hard to move away from the UUID based approach.

> Nice. I could add it to the docs :) . At which point do you get a hold of both filename and path?

During upload, here's an example:

```haskell
module Web.Controller.UploadedFiles where

import Web.Controller.Prelude
import Web.View.UploadedFiles.Show

import qualified Network.Wai.Parse as Wai

instance Controller UploadedFilesController where
    action CreateUploadedFileAction = do
        uploadedFile <- withTransaction do 
            uploadedFile <-
                    newRecord @UploadedFile
                    |> setJust #userId currentUserId
                    |> createRecord

            let file = fileOrNothing "file" |> fromMaybe (error "no file given")

            let options :: StoreFileOptions = def
                    { directory = "uploaded_files"
                    , contentDisposition = contentDispositionAttachmentAndFileName
                    }

            storedFile <- storeFileWithOptions file options

            signedUrl <- createTemporaryDownloadUrl storedFile

            uploadedFile
                |> set #publicUrl (signedUrl |> get #url)
                |> set #publicUrlExpiredAt (signedUrl |> get #expiredAt)
                |> set #objectPath (storedFile |> get #path)
                |> set #fileName (cs $ get #fileName file)
                |> set #contentType (cs $ Wai.fileContentType file)
                |> updateRecord

        render ShowView { .. }
```

Reply:
> but the storage system right now is hard to move away from the UUID based approach.

Yeah, I'm ok with keeping the UUID. So I think we can extend [StoreFileOptions](https://github.com/digitallyinduced/ihp/blob/248cb16b361adb4e8838b33a496f7c2be09541de/IHP/FileStorage/Types.hs#L24C5-L28) to have a `fileName` property. And then I could call `storeFileFromPath` with the filename I'd like (which would be the original UUID). Sounds ok?

Reply:
I think only `storeFileWithOptions` will have to change slightly (with no BC changes). I'll create a PR to confirm

Reply:
https://github.com/digitallyinduced/ihp/pull/1688

Reply:
@mpscholten in your code example above, you are doing a few nice things that I think I could update the docs with. But I to ensure the approach:

1. If we'd have the `attachement` table it means we'd want to separate the file upload from the record build (they are [one](https://github.com/digitallyinduced/ihp/blob/f2f2af6220bdbe04df21e34b13ad561f6969d090/Guide/file-storage.markdown?plain=1#L202) now). So, like in your example we'd start with `withTransaction`, upload the file, create an `Attachement` record, and then create a `Company` record with the `logoUrl` populated.
1. We tell folks to have the `logo_url` as a [Maybe value](https://github.com/digitallyinduced/ihp/blob/f2f2af6220bdbe04df21e34b13ad561f6969d090/Guide/file-storage.markdown?plain=1#L189-L193). But what if the logoUrl should be a required field? With the above approach, we can have logoUrl be a `Text` and not `Maybe Text` -- which is better modeling the fact it's required.

Reply:
> The approach in the docs is a bit simpler.

Oh yeah, I intend to keep the simpler option. But also provide the more advanced one.